data-products/src/main/scala/org/sunbird/analytics/model/report/AssessmentCorrectionModel.scala

```diff
@@ -0,0 +1,82 @@
+
+  implicit val className: String = "org.sunbird.analytics.model.report.AssessmentCorrectionModel"
+  override def name: String = "AssessmentCorrectionModel"
+
+  val redisIndex = AppConf.getConfig("redis.db")
+  val redisHost = AppConf.getConfig("redis.host")
+  val redisPort = AppConf.getConfig("redis.port")
+  val assessEvent = "ASSESS"
+  val contentType = "SelfAssess"
+
+  override def preProcess(data: RDD[String], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[V3Event] = {
+    data.map(f => JSONUtils.deserialize[V3Event](f)).filter(f => null != f.eid && f.eid.equals(assessEvent))
+  }
+
+  override def algorithm(events: RDD[V3Event], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[AssessOutputEvent] = {
+    implicit val sqlContext = new SQLContext(sc)
+
+    val groupedDF = getAssessEventData(events)
+    val redisData = loadRedisData()
+
+    val joinedDF = groupedDF.join(redisData, groupedDF.col("contentid") === redisData.col("identifier"), "inner")
+      .withColumn("attemptId",
+        md5(concat(groupedDF.col("contentid"), groupedDF.col("courseid"),
+          groupedDF.col("batchid"), groupedDF.col("userid"), lit(System.currentTimeMillis()))))
+      .select(groupedDF.col("*"), col("attemptId"))
+
+    val outputData = joinedDF.rdd.map{f =>
+      val assessmentTs: Long = System.currentTimeMillis()
+      AssessOutputEvent(assessmentTs, f.getString(1), f.getString(2), f.getString(3), f.getString(5), f.getString(0), f.getList(4) )
+    }
+    outputData
+  }
+
+  override def postProcess(events: RDD[AssessOutputEvent], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[AssessOutputEvent] = {
+    events
+  }
+
+  def loadRedisData()(implicit sc: SparkContext, sqlContext: SQLContext) = {
+    import sqlContext.implicits._
+    val redisData = sc.fromRedisKV("*")
+    val selfAssessFiltered = redisData
+      .map(f => (f._1, JSONUtils.deserialize[Map[String, AnyRef]](f._2)))
+      .filter(f => contentType.equals(f._2.getOrElse("contentType", "")))
+      .map(f => SelfAssessData(f._1, f._2.getOrElse("contentType", "").asInstanceOf[String]))
+      .toDF()
+    selfAssessFiltered
+  }
+
+  def getAssessEventData(events: RDD[V3Event])(implicit sqlContext: SQLContext): DataFrame = {
+    import sqlContext.implicits._
+    val df = events.map{f =>
+      val userId = if (f.actor.`type`.equals("User")) f.actor.id else ""
+      val cData = f.context.cdata.getOrElse(List())
+      val batchList = cData.filter(f => f.`type`.equalsIgnoreCase("batch")).map(f => f.id)
+      val batchId = if (!batchList.isEmpty) batchList.head else ""
+      val courseList = cData.filter(f => f.`type`.equalsIgnoreCase("course")).map(f => f.id)
+      val courseId = if (!courseList.isEmpty) courseList.head else ""
```

Comment:
I think the type comes in as CourseBatch and Course. Can we check the data once? to check how mobile and portal are sending the cdata?